Drop the dependency on kitchen.

bowlofeggs opened the pull request:
Bodhi had been using kitchen to do a few different things, all of
which were either not useful or silly:

0. Encode strs into strs.
1. Encode strs into bytes. This is silly because we can do that
   just as easily and more correctly with the standard library
   (more correctly because kitchen uses the replace method on
   errors - we would be much better off to know about the errors
   so we can fix them).
2. Iterate on iterables, or turn a non-iterable into a single item
   iterable.

Each of these were replaced by using the standard library.

Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs committed:
Drop the dependency on kitchen.

Bodhi had been using kitchen to do a few different things, all of
which were either not useful or silly:

0. Encode strs into strs.
1. Encode strs into bytes. This is silly because we can do that
   just as easily and more correctly with the standard library
   (more correctly because kitchen uses the replace method on
   errors - we would be much better off to know about the errors
   so we can fix them).
2. Iterate on iterables, or turn a non-iterable into a single item
   iterable.

Each of these were replaced by using the standard library.

Co-Authored-By: Miro Hrončok <miro@hroncok.cz>
Signed-off-by: Randy Barlow <randy@electronsweatshop.com>

bowlofeggs commented:
@hroncok I got you an early Christmas present.

hroncok commented on bodhi/server/mail.py:
```diff
@@ -412,14 +401,14 @@ def send_mail(from_addr, to_addr, subject, body_text, headers=None):
+    body_text = body_text.encode('utf-8')
 
-    msg = [b'From: %s' % to_bytes(from_addr), b'To: %s' % to_bytes(to_addr)]
+    msg = [b'From: %s' % from_addr.encode('utf-8'), b'To: %s' % to_addr.encode('utf-8')]
```
```suggestion
    msg = [f'From: {from_addr}'.encode('utf-8'), f'To: {to_addr}'.encode('utf-8')]
```

hroncok replied:
Also, it might be easier to encode the message once it's joined.

hroncok replied:
Something like https://github.com/bowlofeggs/bodhi/pull/2

hroncok committed:
Stop encoding e-mail line by line

Instead of encoding every line of a message in send_mail(),
we only work with unicode strings and pass such body to
_send_mail(), not encoding it.

We encode the body as we pass it to SMTP.sendmail();
that is needed because otherwise it would be encoded as ASCII.

Now we only convert to bytes at the IO boundary 🧱

Signed-off-by: Miro Hrončok <miro@hroncok.cz>

Zlopez reviewed (approved):
LGTM

bowlofeggs commented:
@hroncok I am going to apply this patch on top of yours to fix a couple of tests that expected the old behavior:

```
diff --git a/bodhi/tests/server/consumers/test_composer.py b/bodhi/tests/server/consumers/test_composer.py
index 6394bb00..dbcb64db 100644
--- a/bodhi/tests/server/consumers/test_composer.py
+++ b/bodhi/tests/server/consumers/test_composer.py
@@ -559,7 +559,7 @@ References:
                                 mock.ANY)
         self.assertEqual(len(mail.mock_calls), 2)
         body = mail.mock_calls[1][1][2]
-        self.assertTrue(body.decode('utf-8').startswith(
+        self.assertTrue(body.startswith(
             ('From: updates@fedoraproject.org\r\nTo: %s\r\nX-Bodhi: fedoraproject.org\r\nSubject: '
              'Fedora 17 updates-testing report\r\n\r\nThe following builds have been pushed to '
              'Fedora 17 updates-testing\n\n    bodhi-2.0-1.fc17\n\nDetails about builds:\n\n\n====='
diff --git a/bodhi/tests/server/test_mail.py b/bodhi/tests/server/test_mail.py
index 9fbd6552..af510456 100644
--- a/bodhi/tests/server/test_mail.py
+++ b/bodhi/tests/server/test_mail.py
@@ -351,7 +351,7 @@ class Test_SendMail(unittest.TestCase):
         mail._send_mail('archer@spies.com', 'lana@spies.com', 'hi')
 
         SMTP.assert_called_once_with('smtp.fp.o')
-        smtp.sendmail.assert_called_once_with('archer@spies.com', ['lana@spies.com'], 'hi')
+        smtp.sendmail.assert_called_once_with('archer@spies.com', ['lana@spies.com'], b'hi')
         warning.assert_called_once_with(
             '"recipient refused" for \'lana@spies.com\', {}'.format(
                 repr(smtp.sendmail.side_effect)))
```

bowlofeggs commented:
The integration tests failed due to hitting a CI timeout. I've opened https://github.com/fedora-infra/bodhi/pull/3155 to raise that timeout.

I ran the CI tests locally on this patch and they all pass, so I'm going to merge this.

bowlofeggs commented:
This patch is planned for inclusion in the upcoming 4.0.0 release: #3221

bowlofeggs commented:
Bodhi 4.0.0 beta is built and deployed to staging:

https://copr.fedorainfracloud.org/coprs/bowlofeggs/bodhi-pre-release
https://bodhi.stg.fedoraproject.org/
